File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.java
class Solution {
    int answer = 0;
    public void dfs(int idx, int sum, int[] numbers, int target){
        if(idx == numbers.length){
            if(sum==target) answer++;
            return;
        }else{
            dfs(idx+1, sum+numbers[idx], numbers, target);
            dfs(idx+1, sum-numbers[idx], numbers, target);
        }
    }
    public int solution(int[] numbers, int target) {
        dfs(0,0,numbers,target);
        return answer;
    }
}

File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.cpp
#include <string>
#include <vector>
#include <queue>

using namespace std;

struct Node{
    int idx;
    int sum;
};

int solution(vector<int> numbers, int target) {
    int answer = 0;
    queue<Node> q;
    
    q.push({0,0});
    while(!q.empty()){
        //큐에서 가장 앞에 애를 꺼냄 
        Node cur = q.front();
        q.pop();
        
        if(cur.idx==numbers.size()){
            if(cur.sum==target) answer++;
            continue;
        }
        
        q.push({cur.idx+1, cur.sum+numbers[cur.idx]});
        q.push({cur.idx+1, cur.sum-numbers[cur.idx]});
    }
    
    return answer;
}